src/dynmem.hpp

```diff
@@ -15,110 +15,191 @@
 #endif
 #endif
 
-class MemoryManager {
+#if !defined(CUFFT) && defined(CUDA_TRAD)
+#error "CUFFT needs to be enabled if CUDA_TRAD is used"
+#endif
+
+#ifdef CUDA_TRAD
+enum MemLoc{
+    CURR_HOST,
+    CURR_DEV
+};
+#endif /* CUDA_TRAD */
+
+template <typename T>
+struct MemNode {
+  T *ptr_h;
+#ifdef CUFFT
+  T *ptr_d;
+#ifdef CUDA_TRAD
+  mutable MemLoc currLoc;
+#endif /* CUDA_TRAD */
+#endif /* CUFFT */
+};
+
+template <typename T> class MemoryManager_ {
     std::mutex mutex;
-    std::map<size_t, std::stack<void*> > map;
+    std::map<size_t, std::stack<T> > map;
 
 public:
-    void *get(size_t size) {
+    bool get(T *node, size_t size) {
         std::lock_guard<std::mutex> guard(mutex);
         auto &stack = map[size];
-        void *ptr = nullptr;
         if (!stack.empty()) {
-            ptr = stack.top();
+            *node = stack.top();
             stack.pop();
+            return true;
         }
-        return ptr;
+        return false;
     }
-    void put(void *ptr, size_t size) {
+    void put(T *node, size_t size) {
         std::lock_guard<std::mutex> guard(mutex);
-        map[size].push(ptr);
+        map[size].push(*node);
     }
 };
 
+typedef MemoryManager_<struct MemNode<void>> MemoryManager;
+
 template <typename T> class DynMem_ {
   private:
-    T *ptr_h = nullptr;
-#ifdef CUFFT
-    T *ptr_d = nullptr;
+    struct MemNode<T> mem;
     static MemoryManager mmng;
-#endif
   public:
     typedef T value_type;
     const size_t num_elem;
 
     DynMem_(size_t num_elem) : num_elem(num_elem)
     {
+        if (!mmng.get(reinterpret_cast<struct MemNode<void>*>(&mem), num_elem))
```

Comment:
Add a new private methods `allocMem` and call it here. There will be three version of this method based on the #ifdefs.